sql/log.cc

```diff
@@ -12863,3 +12936,215 @@ void wsrep_register_binlog_handler(THD *thd, bool trx)
+  // space for Gtid_log_event
+  required_size+= LOG_EVENT_HEADER_LEN + Gtid_log_event::max_data_length +
+                  BINLOG_CHECKSUM_LEN;
+  // space for Empty_event
+  required_size+= LOG_EVENT_HEADER_LEN + BINLOG_CHECKSUM_LEN;
+
+  if (DBUG_IF("simulate_required_size_too_big") ||
+      required_size > m_cache_data->file_reserved_bytes())
+  {
+    sql_print_information("Could not rename binlog cache to binlog, "
+                          "require %llu bytes but only %llu bytes reserved.",
+                          required_size, m_cache_data->file_reserved_bytes());
+    return false;
+  }
+  char temp_name[FN_REFLEN];
+  File read_fd= -1;
+  File new_log_fd= -1;
+  bool ret= false;
+  size_t copy_len= 0;
+  uchar buf[IO_SIZE];
+
+  m_cache_data->tmp_file_name(temp_name);
+
+  read_fd= mysql_file_open(key_file_binlog, mysql_bin_log.get_log_fname(),
+                           O_RDONLY | O_BINARY | O_SHARE, MYF(MY_WME));
+  /* Create fd for the cache file as a new binlog file fd */
+  new_log_fd= mysql_file_open(key_file_binlog, temp_name,
+                              O_BINARY | O_CLOEXEC | O_WRONLY, MYF(MY_WME));
+  if (read_fd == -1 || new_log_fd == -1)
+    goto end;
+
+  /* copy the content of binlog file into the binlog cache's header */
+  while (copy_len < binlog_size)
+  {
+    int read_len= (int) mysql_file_read(read_fd, buf, IO_SIZE, MYF(MY_WME));
+    if (read_len < 0 ||
+        mysql_file_write(new_log_fd, buf, read_len,
+                         MYF(MY_WME | MY_NABP | MY_WAIT_IF_FULL)))
+      goto end;
+
+    copy_len+= read_len;
+  }
+
+  if (DBUG_IF("simulate_write_binlog_cache_head_error") ||
+      mysql_file_sync(new_log_fd, MYF(MY_WME)))
+    goto end;
+
+  my_delete(mysql_bin_log.get_log_fname(), MYF(0));
+
+  if (DBUG_IF("simulate_rename_binlog_cache_to_binlog_error") ||
+      my_rename(temp_name, mysql_bin_log.get_log_fname(), MYF(MY_WME)))
+  {
+    /*
+      Since the binlog file is deleted, it should return true if error
+      happens. Thus the index file will be reverted later.
+    */
+    ret= true;
+    goto end;
+  }
+
+  sql_print_information("Renamed binlog cache to binlog %s",
+                        mysql_bin_log.get_log_fname());
+  mysql_file_close(mysql_bin_log.log_file.file, MYF(MY_WME));
+  mysql_bin_log.log_file.file= new_log_fd;
+  new_log_fd= -1;
+  m_replaced= true;
+end:
+  if (read_fd != -1)
+    mysql_file_close(read_fd, MYF(MY_WME));
+  if (new_log_fd != -1)
+    mysql_file_close(new_log_fd, MYF(MY_WME));
+  return ret;
+}
+
+bool Binlog_cache_free_flush::complete()
+{
+  size_t begin_pos = my_b_tell(&mysql_bin_log.log_file);
+  size_t empty_size= m_cache_data->file_reserved_bytes() - begin_pos
+                     - LOG_EVENT_HEADER_LEN;
+
+  if (binlog_checksum_options != BINLOG_CHECKSUM_ALG_OFF)
+    empty_size-= BINLOG_CHECKSUM_LEN;
+
+  Empty_log_event empty_ev(m_entry->thd, (int) empty_size);
+  if (mysql_bin_log.write_event(&empty_ev) ||
```

Comment:
Look the code in Gtid_log_event::write(), gtid log event also has variable length. with/o XID, single/multiple engine, it has different length.  I also thought about using a Query_log_event instead of a empty event, but that is also hack and a little bit complex than an empty event.

Reply:
You should be able to reserve the maximum size of a GTID event (which you already pre-calculate into `Gtid_log_event::max_data_length`) and pad the unused data with `0`s. The padded `0`s should just be treated as reserve bytes (see the Gtid_log_event class comment for more details), provided the `event_length` field is consistent with the maximum size. 

And another thought comes to mind, I think the GTID event should be written before the log rotation/renaming happens. That way, the rotate event never points to a corrupt log.

Reply:
You are right, Gtid_log_event has fixed length. thanks for the information. 
As you said Gtid_list_log_event's length may change, and 
checkpoint event's length may change as well in the situation that the suffix number is larger than 999999. 
So we cannot suppose it has fixed length most time. but that is not very often to change.  We probable can detect the change when rotating binlog and updated the reserved length for binlog caches.

In that way, we can remove the empty event. But there is still some extra work to do. Binlog cache is aligned with IO_SIZE(4K), thus it will waste the first reserved size bytes of memory, or we need to add our own buffer instead of IO_CACHE's buffer.  I will try to remove the empty event.

Reply:
Hi @bnestere, I removed the empty event, in the latest patch. The the space left is padded into Gtid event.
It still reserved 4K(IO_SIZE) space for easily handle the alignment. 

I tried to write gtid event earlier but that is a little complex, because the gtid needs to be written into the temp file.
so I think it is better not to spread free_flush related context to more places and  keep the original binlog_cache flush process. However I have no strong opinion on it.

Reply:
> It still reserved 4K(IO_SIZE) space for easily handle the alignment.

Hmm, sorry, why does aligning the cache data to the start of a page help in this case? I don't see a performance benefit (though that doesn't mean there isn't one :innocent:), and it adds complexity in the code, and awkwardness into the event.

> I tried to write gtid event earlier but that is a little complex, because the gtid needs to be written into the temp file. so I think it is better not to spread free_flush related context to more places and keep the original binlog_cache flush process. However I have no strong opinion on it.

Can you write it in `replace_binlog_file()` before doing the actual file replacement? That'd keep the spread minimal while also keeping the state consistent.

Reply:
> Hmm, sorry, why does aligning the cache data to the start of a page help in this case? I don't see a performance benefit (though that doesn't mean there isn't one 😇), and it adds complexity in the code, and awkwardness into the event.
look the code below
```
  if (type == WRITE_CACHE)
    info->write_end=
      info->buffer+info->buffer_length- (seek_offset & (IO_SIZE-1));
```
the buffer of IO_CACHE is 4K aligned. Since we reserved e.g. 400 bytes, the seek_offset will be 400, thus only 4k-400B can be used, though 4K is allocated. It has no performance improvement, it just eliminates memory waste.

Reply:
Ok, I suppose it would be a trade-off either way. Though keeping with the aligned approach, I'd suggest adding some additional logic when setting `binlog_cache_data::m_file_reserved_bytes` to calculate the size of the header bytes to see if it exceeds `IO_SIZE` (if the `Gtid_list_log_event` is very big) and set it to the next multiple of `IO_SIZE` if so.

Reply:
ok, In the original patch, it records the header size of the new binlog file when rotating. and use it(aligned to IO_SIZE) as the reserved space. I will add it back. I  had  removed the logic just because I thought  mariadb's gtid_list_log_event would not be  big.